pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,470 @@
+			}
+		} else if toIP.To16() != nil && ipv6Socket != nil {
+			if err := m.sendPing(ipv6Socket, toIP); err != nil {
+				klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IPv6", "IP", toIP)
+			}
+		} else {
+			klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IP family", "IP", toIP)
+		}
+	}
+	klog.V(4).InfoS("Done pinging all Nodes")
+}
+
+// Run starts the NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+
+	<-stopCh
+}
+
+// monitorLoop is the main loop to monitor the latency of the Node.
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	klog.V(4).InfoS("NodeLatencyMonitor is running")
+	// Low level goroutine to handle ping loop
+	var ticker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if ticker != nil {
+			ticker.Stop() // Stop the current ticker
+		}
+		ticker = time.NewTicker(interval)
+		tickerCh = ticker.C
+	}
+
+	klog.V(4).InfoS("NodeLatencyMonitor is running2")
+	wg := sync.WaitGroup{}
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+		case <-stopCh:
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// latencyConfig changed
+				updateTicker(m.latencyConfig.Interval)
+
+				// If the recvPing socket is closed,
+				// recreate it if it is closed(CRD is deleted).
+				if ipv4Socket == nil && m.isIPv4Enabled {
+					// Create a new socket for IPv4 when it is IPv4-only
+					ipv4Socket, err = icmp.ListenPacket(IPv4ProtocolICMPRaw, "0.0.0.0")
+					if err != nil {
+						klog.ErrorS(err, "Failed to create ICMP socket for IPv4")
+						return
+					}
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						m.recvPing(ipv4Socket, true)
+					}()
+				}
+				if ipv6Socket == nil && m.isIPv6Enabled {
+					// Create a new socket for IPv6 when it is IPv6-only
+					ipv6Socket, err = icmp.ListenPacket(IPv6ProtocolICMPRaw, "::")
+					if err != nil {
+						klog.ErrorS(err, "Failed to create ICMP socket for IPv6")
+						return
+					}
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						m.recvPing(ipv6Socket, false)
+					}()
+				}
+			} else {
+				// latencyConfig deleted
+				if ticker != nil {
+					ticker.Stop()
+					ticker = nil
```

Comment:
should `tickerCh` be reset? otherwise it would ping once if the ticker already fires.

Reply:
Reset is better and I will update it.